portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -183,19 +203,65 @@ private String getUploadedFileUrl(HttpServletRequest request, HttpServletRespons
     if (blobInfo.getContentType().contains("image") == false) {
       response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Only images allowed!");
       blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    return blobKey;
+  }
+
+  /**
+   * Blobstore stores files as binary data. This function retrieves the binary data stored at the
+   * BlobKey parameter.
+   */
+  private byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
```

Comment:
This code cannot successfully read images larger than MAX_BLOB_FETCH_SIZE.

MAX_BLOB_FETCH_SIZE is meant to be an upper limit for how much you can fetch with each `fetchData` call, not an upper limit for the blob size.

It looks like you need to call another method first to retrieve the image size: http://bpossolo.blogspot.com/2012/04/method-for-fully-reading-blob-from.html

Then you can figure out how many `fetchData` calls you need to retrieve the entire image.

Reply:
But what do you mean? It fetches multiple times a number of bytes equal to MAX_BLOB_FETCH_SIZE, until it reaches the end of the byte stream, and it writes the data in outputBytes.